core/src/main/java/org/apache/calcite/sql/SqlSessionTableFunction.java

```diff
@@ -0,0 +1,71 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.type.SqlTypeUtil;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+/**
+ * SqlSessionTableFunction implements an operator for per-key sessionization. It allows
+ * four parameters:
+ * 1. a table.
+ * 2. a descriptor to provide a watermarked column name from the input table.
+ * 3. a descriptor to provide a column as key, on which sessionization will be applied.
+ * 4. an interval parameter to specify a inactive activity gap to break sessions.
+ */
+public class SqlSessionTableFunction extends SqlWindowTableFunction {
+  public SqlSessionTableFunction() {
+    super(SqlKind.SESSION.name());
+  }
+
+  @Override public SqlOperandCountRange getOperandCountRange() {
+    return SqlOperandCountRanges.of(4);
+  }
+
+  @Override public boolean checkOperandTypes(SqlCallBinding callBinding,
+      boolean throwOnFailure) {
+    final SqlNode operand0 = callBinding.operand(0);
+    final SqlValidator validator = callBinding.getValidator();
+    final RelDataType type = validator.getValidatedNodeType(operand0);
+    if (type.getSqlTypeName() != SqlTypeName.ROW) {
+      return throwValidationSignatureErrorOrReturnFalse(callBinding, throwOnFailure);
+    }
+    final SqlNode operand1 = callBinding.operand(1);
+    if (operand1.getKind() != SqlKind.DESCRIPTOR) {
+      return throwValidationSignatureErrorOrReturnFalse(callBinding, throwOnFailure);
+    }
+    validateColumnNames(validator, type.getFieldNames(), ((SqlCall) operand1).getOperandList());
+    final SqlNode operand2 = callBinding.operand(2);
+    if (operand2.getKind() != SqlKind.DESCRIPTOR) {
+      return throwValidationSignatureErrorOrReturnFalse(callBinding, throwOnFailure);
+    }
+    validateColumnNames(validator, type.getFieldNames(), ((SqlCall) operand2).getOperandList());
+    final RelDataType type3 = validator.getValidatedNodeType(callBinding.operand(3));
+    if (!SqlTypeUtil.isInterval(type3)) {
+      return throwValidationSignatureErrorOrReturnFalse(callBinding, throwOnFailure);
+    }
+    return true;
+  }
+
+  @Override public String getAllowedSignatures(String opNameToUse) {
+    return getName() + "(TABLE table_name, DESCRIPTOR(col), "
+        + "DESCRIPTOR(col), datetime interval)";
```

Comment:
Why we must have 2 `DESCRIPTOR` here ?

Reply:
For session, usually common usage is per key. E.g. per user session. That's what the second descriptor used for: offer a session key.

The group window function does not have this problem because that style of windowing happens at GROUP BY, where users can put separate GROUP BY key as session key.

Reply:
I mean why couldn't we have one `DESCRIPTOR` with 2 columns ?

Reply:
Because descriptor only provide column names so far. If there is only one descriptor in which there are two column names, it will have a problem to know which one is the timestamp column that apply windowing upon and which one is the session key(note that session could be timestamp type as well). Of course a workaround is to define, say: the first column name is for windowing, and the second column is for session key. But this is error-prone, both from user usage and documentation: users have to use the right column name on the right position and documentation needs to explain the meaning of column position.

Two descriptors is better as each one has its clear semantic and require one column only. 

So I think it's a good API design than single descriptor.

Reply:
I don't think so, use two descriptors also has such concerns, people would also ask "which column is timestamp column"

Reply:
It's ok that you don't think that benefit holds. Even if so, then seems to me that there is no difference between one descriptor and two descriptors. What do you think?

Reply:
The `descriptor` keyword itself is too verbose, why we must have that ? Couldn't we just put a column name there directly ?

Reply:
In fact, it is a feature defined in SQL standard 2016: Polymorphic table functions(https://issues.apache.org/jira/browse/CALCITE-2270). `DESCRIPTOR` is designed to specify columns explicitly in table functions. That was one of the purpose  it was created for.

Reply:
Thanks for the explanations, personally i think one `DESCRIPTOR` is more concise, i'm wondering what other people think ?

Reply:
Can the session key be multiple keys ?

Reply:
This is a good point. In my implementation, it is single key only (cause that is the typical usage). But compound key is feasible and valid use case.

If there is a need to support compound key, then I think two descriptors will be better.

Reply:
Yes, two descriptor is more feasible to support multiple session keys, we better add a test in  SqlToRelConverterTest to support multiple session keys.

Reply:
Added a `testTableFunctionSessionCompoundSessionKey` in `SqlToRelConverterTest` for verifying that generated plan is correct when there are multiple session key.

Reply:
Another reason why we use descriptor than just a column name: https://issues.apache.org/jira/browse/CALCITE-3964

Basically there is a missing feature of `DESCRIPTOR`: an optional data type, which can be used to do type validation. It's useful for event timestamp windowing because such windowing should be apply on TIMESTAMP column. `DESCRIPTOR` will be handy to validate if user query applies windowing on a right column.

Reply:
From the paper, the DESCRIPTOR only appears in the function definition, that is, it is in the DDL, here we are not creating a function, but use it directly, i'm not sure if we should use the descriptor, as for the first argument, i would fix it as a cursor.